### ISSUE TITLE
location/velocity/acceleration cache isn't being invalidated when integrating with the manager

### ISSUE THREAD

**Issue by carmichaelong** [opensim-org/opensim-core#1369]
Looked at this with @aseth1. Simple way to reproduce:

In `testPoints`, there is a test that uses a `TimeStepper` and compares the location, velocity, and acceleration of the center of mass of the pendulum between the `Station` method and `MobilizedBody` methods directly. This test passes as is, but does not pass if the integration is done by a `Manager` instead. Location, velocity, and acceleration match for the first (initial) time, but then all three fail to match on the immediate next time step.

**Comment by jenhicks** [comment]
@carmichaelong Thanks for reporting. Sounds like something we should certainly fix for 4.0. Is it affecting your work? If so, we will prioritize.

**Comment by aseth1** [comment]
I do not believe this is a bug in the Frame or Point classes and if we use the `TimeStepper` directly and only initialize(State) once, these lazy cache entries work as expected.

Looking a bit deeper, the `TimeStepper::initialize(State)` calls the Integrator's `initialize`, which makes a copy of the state passed in for its internal use, but unlike typical copies, it seems that the default system cache variables are not invalidated. The cache variables seem to be untouched during `Manager::doIntegration` (which invokes the `TimeStepper`) and it updates the passed in State reference to now refer to the integrator's internal copy (with now stale cache). 

If we explicitly make a copy of the internal state (instead of a getting a reference) we do get the cache to be invalidated and subsequent calls to the Frame and Point methods are forced to to compute their transform and location.

Unlike `GeometryPath`'s path and other lazily cached variables we are using (like StateVariable derivatives), the transform and location cache's of `Frame` and `Point` are unused during the evaluation of derivatives and the process of integration (there is no reason for the lazy cache variables to be recomputed), and so these cache entries appear to be untouched - not even invalidated.

@sherm1 do you have any comments/concerns/advice regarding this issue? Is it valid to create a new TimeStepper with the same Integrator repeatedly?  Thanks!

**Comment by aseth1** [comment]
@jenhicks I agree this is a bug, but it isn't completely clear if it is because of a misuse of the TimeStepper (e.g. calling `ts.initialize(s)` more than once) or if there is a bug within `TimeStepper` or `Integrator` that should be invalidating the cache variables and it isn't. I hope @sherm1 can chime in.

**Comment by sherm1** [comment]
Hi, @aseth1. Reading this my first thought was about interpolation. The integrator maintains two internal states, one is the "advanced" state that is the actual trajectory, and one is used for interpolation. If you use `getState()` the returned reference switches between those two. There is also `getAdvancedState()` that always returns the trajectory one. Is it possible that Manager is thinking that there is a single internal state in the integrator and is therefore sometimes looking at a reference to the wrong internal state? The advanced state is always valid (though possibly a few steps into the future). The interpolated state is just a temp and will have junk in it except when it is in use.

**Comment by aseth1** [comment]
Thanks @sherm1. That sounds like an easy enough fix to replace  `getState()` with `getAdvancedState()`. @carmichaelong in your test harness that you have setup, can you try replacing and report what happens? Thank you for investigating.

**Comment by sherm1** [comment]
Caveat: that is only what first came to mind -- I'm not at all sure that is the actual problem!

**Comment by jenhicks** [comment]
Related to #949

**Comment by carmichaelong** [comment]
I tested this in a smaller case (that doesn't depend on the `Manager` at all). The setup is as follows:

```
SimTK::RungeKuttaMersonIntegrator integrator(pendulum.getSystem());
SimTK::TimeStepper ts(pendulum.getSystem(), integrator);
ts.initialize(state);
...
for (int i = 1; i <= n; ++i) {
        //ts.initialize(state);
        ts.stepTo(i*dt);
        state = ts.getState();
        ...
```
The code as shown works. If the line that is commented is added in, this breaks causes an issue. In response to `getState()` versus `getAdvancedState()`, the `TimeStepper` only has a method for the first. So, I replaced the last line with `state = ts.updIntegrator().getAdvancedState();`. This does not fix the issue.

**Comment by sherm1** [comment]
Why would the Manager initialize the time stepper every step?

**Comment by carmichaelong** [comment]
The Manager creates a `TimeStepper` each time `integrate()` is called (see [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/Manager/Manager.cpp#L736)). This means the code above isn't quite the same, but my testing of creating a new `TimeStepper` in each loop had the same effect. Note that the Integrator isn't created each time by the Manager.

**Comment by aseth1** [comment]
@carmichaelong in the typical use case, the `Manager` does not create the `TimeStepper` every timestep since it is supposed to integrate the entire interval specified. But, `Manager` does create a `TimeStepper` every `Manager::doIntegration(s)` which is called only once by `Manager::integrate(s)`. It should be possible to avoid the creation of a `TimeStepper`  every `integrate` and and initialize once to mitigate the problem we are seeing when using the Manager like a `TimeStepper` to repeatedly integrate short intervals sequentially to build up a longer period.

But before we do that, these two lines [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/Manager/Manager.cpp#L811):
```cpp     
finalize(_integ->updAdvancedState() );
s = _integ->getState(); 
```
seem troubling, since according to @sherm1's comments, it would mean that the advanced state is updated while we are returning a potentially stale state from `getState()`.

**Comment by carmichaelong** [comment]
I agree that this is not an issue in the typical use case, just when the `Manager` is used as if it were a `TimeStepper` that could be used multiple times. Sounds like we do indeed want the `Manager` to have this functionality. I started on some fixes so that a `TimeStepper` is only created and initalized once, and that the `Manager` throws an Exception if either `setIntegrator` or `setModel` is called after a call to `integrate(s)` has been made.

I'll look into those last two lines as well. @sherm1 did mention one thing:

> The advanced state is always valid (though possibly a few steps into the future).

I don't think I know enough about this to know if that will be an issue or not (or at the very least change some behavior unexpectedly) if `getAdvancedstate()` is used instead. I think this also begs another question: is `Integrator::getAdvancedState()` going to be the same as `TimeStepper::getState()` in all cases? If not, which should one use?

**Comment by sherm1** [comment]
The TimeStepper::getState() isn't likely to be the same state as the integrator's getAdvancedState() because it is allowed to return an interpolated state (unless specifically disallowed). In that case a temp State is used for the interpolations until it catches up with the advanced state, which is off into the future.

**Comment by aseth1** [comment]
closed by #1386
